src/query/storages/fuse/src/io/write/inverted_index_writer.rs

```diff
@@ -0,0 +1,408 @@
+
+pub struct InvertedIndexWriter {
+    schema: DataSchema,
+    index_writer: SingleSegmentIndexWriter,
+}
+
+impl InvertedIndexWriter {
+    pub fn try_create(schema: DataSchema) -> Result<InvertedIndexWriter> {
+        let text_field_indexing = TextFieldIndexing::default()
+            .set_tokenizer("en")
+            .set_index_option(IndexRecordOption::WithFreqsAndPositions);
+        let text_options = TextOptions::default().set_indexing_options(text_field_indexing);
+
+        let mut schema_builder = Schema::builder();
+        let mut index_fields = Vec::with_capacity(schema.fields.len());
+        for field in &schema.fields {
+            if field.data_type().remove_nullable() != DataType::String {
+                return Err(ErrorCode::IllegalDataType(format!(
+                    "inverted index only support String type, but got {}",
+                    field.data_type()
+                )));
+            }
+            let index_field = schema_builder.add_text_field(field.name(), text_options.clone());
+            index_fields.push(index_field);
+        }
+        let index_schema = schema_builder.build();
+
+        let index_settings = IndexSettings {
+            sort_by_field: None,
+            ..Default::default()
+        };
+        let tokenizer_manager = TokenizerManager::new();
+        tokenizer_manager.register(
+            "en",
+            TextAnalyzer::builder(SimpleTokenizer::default())
+                .filter(RemoveLongFilter::limit(40))
+                .filter(LowerCaser)
+                .filter(Stemmer::new(Language::English))
+                .build(),
+        );
+
+        let index_builder = IndexBuilder::new()
+            .settings(index_settings)
+            .schema(index_schema.clone())
+            .tokenizers(tokenizer_manager.clone());
+
+        let ram_directory = RamDirectory::create();
+        let index_writer = index_builder.single_segment_index_writer(ram_directory, 50_000_000)?;
+
+        Ok(Self {
+            schema,
+            index_writer,
+        })
+    }
+
+    pub fn add_block(&mut self, block: DataBlock) -> Result<()> {
+        if block.num_columns() != self.schema.num_fields() {
+            return Err(ErrorCode::TableSchemaMismatch(format!(
+                "Data schema mismatched. Data columns length: {}, schema fields length: {}",
+                block.num_columns(),
+                self.schema.num_fields()
+            )));
+        }
+        for (column, field) in block.columns().iter().zip(self.schema.fields().iter()) {
+            if &column.data_type != field.data_type() {
+                return Err(ErrorCode::TableSchemaMismatch(format!(
+                    "Data schema mismatched (col name: {}). Data column type is {:?}, but schema field type is {:?}",
+                    field.name(),
+                    column.data_type,
+                    field.data_type()
+                )));
+            }
+        }
+
+        for i in 0..block.num_rows() {
```

Comment:
could we filter non-string columns to a new block ?

Reply:
The block is read according to the schema of the index, all columns are of type string, so there is no need to filter them separately.